src/read/macho/dyld_cache.rs

```diff
@@ -0,0 +1,214 @@
+        let images_count = self.header.images_count.get(self.endian);
+        DyldCacheImageIterator {
+            cache: self,
+            images_count,
+            next_image_index: 0,
+            next_image_offset: images_offset,
+        }
+    }
+}
+
+/// An iterator over all the images (dylibs) in the dyld shared cache.
+#[derive(Debug)]
+pub struct DyldCacheImageIterator<'data, 'cache, E = Endianness, R = &'data [u8]>
+where
+    E: Endian,
+    R: ReadRef<'data>,
+{
+    cache: &'cache DyldCache<'data, E, R>,
+    images_count: u32,
+    next_image_index: u32,
+    next_image_offset: u64,
+}
+
+impl<'data, 'cache, E, R> DyldCacheImageIterator<'data, 'cache, E, R>
+where
+    E: Endian,
+    R: ReadRef<'data>,
+{
+    /// Advance the iterator and return the current image.
+    pub fn next(&mut self) -> Result<Option<DyldCacheImage<'data, E, R>>> {
+        if self.next_image_index >= self.images_count {
+            return Ok(None);
+        }
+        self.next_image_index += 1;
+        let data = self.cache.data;
+        let image_info = data
+            .read::<macho::DyldCacheImageInfo<E>>(&mut self.next_image_offset)
+            .read_error("Couldn't read macho::DyldCacheImageInfo")?;
+        Ok(Some(DyldCacheImage {
+            endian: self.cache.endian,
+            is_64: self.cache.is_64,
+            data,
+            first_mapping_address: self.cache.first_mapping_address,
+            image_info,
+        }))
+    }
+}
+
+/// One image (dylib) from inside the dyld shared cache.
+#[derive(Debug)]
+pub struct DyldCacheImage<'data, E = Endianness, R = &'data [u8]>
+where
+    E: Endian,
+    R: ReadRef<'data>,
+{
+    endian: E,
+    is_64: bool,
+    data: R,
+    first_mapping_address: u64,
+    image_info: &'data macho::DyldCacheImageInfo<E>,
+}
+
+impl<'data, E, R> DyldCacheImage<'data, E, R>
+where
+    E: Endian,
+    R: ReadRef<'data>,
+{
+    /// The file system path of this image.
+    pub fn path(&self) -> Result<&'data str> {
+        // The longest path I've seen is 164 bytes long. In theory paths could be longer than 256.
+        const MAX_PATH_LEN: u64 = 256;
+
+        let path_offset = self.image_info.path_file_offset.get(self.endian) as u64;
+        let slice_containing_path = self
+            .data
+            .read_bytes_at(path_offset, MAX_PATH_LEN)
+            .read_error("Couldn't read path")?;
+        let path = Bytes(slice_containing_path).read_string().read_error(
+            "Couldn't read path string (didn't find nul byte within first 256 bytes)",
+        )?;
+        // The path should always be ascii, so from_utf8 should alway succeed.
+        let path = core::str::from_utf8(path).map_err(|_| Error("Path string not valid utf-8"))?;
+        Ok(path)
+    }
+
+    /// The offset in the dyld cache file where this image starts.
+    pub fn offset(&self) -> u64 {
+        self.image_info.address.get(self.endian) - self.first_mapping_address
```

Comment:
Is this assuming that all images are in the first mapping?

Reply:
I'm not sure - this is based on the code in [dsc_iterator.cpp](https://opensource.apple.com/source/dyld/dyld-851.27/dyld3/shared-cache/dsc_iterator.cpp.auto.html), specifically the function `forEachDylibInCache` which calls this offset `firstRegionAddress`. It applies the same offset to all images.

Reply:
I read your implementation and I agree it looks more correct. To make sure it doesn't break anything, I looked at a few dyld shared cache files. The "correct" code is equivalent to the simple code if there is only one mapping, or if all the images are in the first mapping, or if all the mappings containing images have the same value for `mapping.address - mapping.offset`.
It seems like the second option is the case: In the cache files I've checked, all images are in the first mapping. The x86_64 and x86_64h caches on 10.14.6 and on 11.3.1 each have three mappings, and the arm64e cache on 11.3.1 has seven mappings. The mappings have different values for `mapping.address - mapping.offset`, but that's fine since all images are in the first mapping.
So, in summary, what you have looks great and shouldn't break anything.